src/components/TradeInput.tsx

```diff
@@ -9,111 +9,473 @@ import {
+    onOpen()
+  }, [onOpen])
+
+  const handleBuyAssetClick = useCallback(() => {
+    setAssetSelectType(AssetType.BUY)
+    onOpen()
+  }, [onOpen])
+
+  const handleAssetSelect = useCallback(
+    (asset: Asset) => {
+      if (assetSelectType === AssetType.BUY) {
+        setValue('buyAssetId', asset.assetId)
+      }
+      if (assetSelectType === AssetType.SELL) {
+        setValue('sellAssetId', asset.assetId)
+      }
+      onClose()
+    },
+    [assetSelectType, setValue, onClose],
+  )
+
+  const handleSwitchAssets = useCallback(() => {
+    const currentSellAsset = sellAsset
+    const currentBuyAsset = buyAsset
+
+    if (
+      !(
+        currentSellAsset &&
+        currentBuyAsset &&
+        sellAssetMarketData &&
+        buyAssetMarketData &&
+        sellAmountCryptoBaseUnit
+      )
+    )
+      return
+
+    // Get current fiat value using market rates only
+    const currentSellAmountCryptoPrecision = fromBaseUnit(
+      sellAmountCryptoBaseUnit,
+      currentSellAsset.precision,
+    )
+    const sellAmountUsd = bnOrZero(currentSellAmountCryptoPrecision).times(
+      sellAssetMarketData.price,
+    )
+
+    // Calculate new sell amount in crypto using market rates, avoiding division by zero
+    const newSellAmountCryptoPrecision = bnOrZero(buyAssetMarketData.price).isZero()
+      ? '0'
+      : sellAmountUsd.div(buyAssetMarketData.price).toString()
+
+    // Convert to base units with new precision
+    const newSellAmountBaseUnit = toBaseUnit(
+      newSellAmountCryptoPrecision,
+      currentBuyAsset.precision,
+    )
+
+    setValue('sellAmountCryptoBaseUnit', newSellAmountBaseUnit)
+
+    setValue('sellAssetId', currentBuyAsset.assetId)
+    setValue('buyAssetId', currentSellAsset.assetId)
+  }, [
+    sellAsset,
+    buyAsset,
+    sellAssetMarketData,
+    buyAssetMarketData,
+    sellAmountCryptoBaseUnit,
+    setValue,
+  ])
+
+  const handleSellAmountChange = useCallback((values: { value: string }) => {
+    setSellAmountInput(values.value)
   }, [])
 
+  const validateDestinationAddress = useCallback(
+    async (value: string) => {
+      if (!value || !buyAssetId) return true
+      const { chainId } = fromAssetId(buyAssetId)
+      if (destinationAddress === value && !errors.destinationAddress) {
+        return true
+      }
+      const isValid = await validateAddress(value, chainId)
+      return isValid || 'Invalid address'
+    },
+    [buyAssetId, destinationAddress, errors.destinationAddress],
+  )
+
+  const validateRefundAddress = useCallback(
+    async (value: string) => {
+      if (!value || !sellAssetId) return true
+      const { chainId } = fromAssetId(sellAssetId)
+      if (refundAddress === value && !errors.refundAddress) {
+        return true
+      }
+      const isValid = await validateAddress(value, chainId)
+      return isValid || 'Invalid address'
+    },
+    [sellAssetId, refundAddress, errors.refundAddress],
+  )
+
+  const destinationAddressRules = useMemo(
```

Comment:
These don't seem to revalidate upon changing assets - it's happy to let me start the transaction with a BTC refund address here:

<img width="356" alt="Screenshot 2025-02-10 at 19 00 40" src="https://github.com/user-attachments/assets/606f7c1a-598f-4690-b241-59b8b00879d3" />

Reply:
18f27bd, though still a *lil* bit flaky https://jam.dev/c/69130c36-2c9b-4e65-91d3-8e120cebbcfd

Reply:
64db3a8543e150f9f240fef7ac66fb833e918ca8 / https://jam.dev/c/725f0461-15d5-42d8-8d82-b5e510735edb